### PR TITLE
config: Add endpoint parameter

### PR DESCRIPTION
**Pull request by zhaohuxing**

### CODE REVIEW

File: request/builder/builder.go [hunk 1]
```diff
@@ -279,8 +281,18 @@ func (qb *Builder) parseRequestURL() error {
 	zone := (*qb.parsedProperties)["zone"]
 	port := strconv.Itoa(config.Port)
 	endpoint := config.Protocol + "://" + config.Host + ":" + port
+	if config.Endpoint != "" {
+		endpoint = config.Endpoint
```
Comment (Xuanwo) [inline request/builder/builder.go]: Maybe we need to parse `endpoint` into `protocol`, `host` and `port` for checking.

Think about following situations:

- endpoint is `192.169.0.10`
- endpoint is `http://192.168.0.10`
- endpoint is `[xxxx:xxxx:xxxx]` (valid ipv6 address)
- endpoint is `https://qingstor.com`
- endpoint is `https://pek3b.qingstor.com`
- endpoint is `https://bucket.pek3b.qingstor.com`
- endpoint is `https://cname-to-qingstor.com`

File: config/config.go [hunk 2]
```diff
@@ -228,10 +229,40 @@ func (c *Config) LoadConfigFromContent(content []byte) (err error) {
...
 	c.InitHTTPClient()
 	return
 }
 
+func (c *Config) parseEndpoint() (err error) {
+	if c.Endpoint == "" {
+		return nil
+	}
+
+	parts := strings.Split(c.Endpoint, "://")
```
Comment (Xuanwo) [inline config/config.go]: Maybe we can use `url.Parse` instead of `strings.Split` here?

Comment (zhaohuxing) [reply config/config.go]: OK

File: config/config.go [hunk 3]
```diff
@@ -224,10 +230,40 @@ func (c *Config) LoadConfigFromContent(content []byte) (err error) {
...
+		return nil
+	}
+
+	u, err := url.Parse(c.Endpoint)
+	if err != nil {
+		return err
+	}
+
+	c.Protocol = u.Scheme
+	parts := strings.Split(u.Host, ":")
```
Comment (Xuanwo) [inline config/config.go]: Maybe we can use `u.Hostname()` and `u.Port()` instead?

File: config/config_test.go [hunk 4]
```diff
@@ -102,3 +102,14 @@ func TestNew(t *testing.T) {
 	assert.Equal(t, "https", config.Protocol)
 	assert.Equal(t, "qingstor.com", config.Host)
 }
+
+func TestParseEndpoint(t *testing.T) {
+	c := Config{
+		Endpoint: "https://qingstor.com:443",
```
Comment (Xuanwo) [inline config/config_test.go]: Use table tests here could be better, we need to test:

```
https://qingstor.com:443
qingstor.com
pek3b.qingstor.com
http://pek3b.qingstor.com:8080
```

and so on.

File: config/config_test.go [hunk 5]
```diff
@@ -105,21 +105,27 @@ func TestNew(t *testing.T) {
...
+		endpoint string
+		protocol string
+		host     string
+		port     int
 	}{
-		{"https://qingstor.com:443", "qingstor.com"},
-		{"https://pek3b.qingstor.com:8080", "pek3b.qingstor.com"},
+		{"https://qingstor.com:443", "https", "qingstor.com", 443},
+		{"https://pek3b.qingstor.com:8080", "https", "pek3b.qingstor.com", 8080},
+		{"https://qingstor.com", "https", "qingstor.com", 0},
```
Comment (Xuanwo) [inline config/config_test.go:115]: port should be 443

File: config/config_test.go [hunk 6]
```diff
@@ -105,21 +105,27 @@ func TestNew(t *testing.T) {
...
+		{"qingstor.com", "", "", 0},
```
Comment (Xuanwo) [inline config/config_test.go:116]: `{"qingstor.com", "https", "qingstor.com", 443},`

**Review by Xuanwo — APPROVED** [review]

**Pull request merged**